client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transaction/SyncModeTransactionTable.java

```diff
@@ -0,0 +1,238 @@
+import javax.transaction.xa.XAResource;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.impl.operations.CompleteTransactionOperation;
+import org.infinispan.client.hotrod.impl.operations.PrepareTransactionOperation;
+import org.infinispan.client.hotrod.impl.transaction.entry.Modification;
+import org.infinispan.client.hotrod.logging.Log;
+import org.infinispan.client.hotrod.logging.LogFactory;
+import org.infinispan.client.hotrod.transaction.manager.RemoteXid;
+import org.infinispan.commons.CacheException;
+
+/**
+ * A {@link TransactionTable} that registers the {@link RemoteCache} as a {@link Synchronization} in the transaction.
+ * <p>
+ * Only a single {@link Synchronization} is registered even if multiple {@link RemoteCache}s interact with the same
+ * transaction.
+ * <p>
+ * When more than one {@link RemoteCache} is involved in the {@link Transaction}, the prepare, commit and rollback
+ * requests are sent sequential and they are ordered by the {@link RemoteCache}'s name.
+ * <p>
+ * If a {@link RemoteCache} is read-only, the commit/rollback isn't invoked.
+ *
+ * @author Pedro Ruivo
+ * @since 9.3
+ */
+public class SyncModeTransactionTable implements TransactionTable {
+
+   private static final Log log = LogFactory.getLog(SyncModeTransactionTable.class, Log.class);
+   private static final boolean trace = log.isTraceEnabled();
+   private final Map<Transaction, SynchronizationAdapter> registeredTransactions = new ConcurrentHashMap<>();
+   private final UUID uuid = UUID.randomUUID();
+
+   @Override
+   public <K, V> TransactionContext<K, V> enlist(TransactionalRemoteCacheImpl<K, V> txRemoteCache, Transaction tx) {
+      //registers a synchronization if it isn't done yet.
+      SynchronizationAdapter adapter = registeredTransactions.computeIfAbsent(tx, this::createSynchronizationAdapter);
+      //registers the cache.
+      TransactionContext<K, V> context = adapter.registerCache(txRemoteCache);
+      if (trace) {
+         log.tracef("Xid=%s retrieving context: %s", adapter.xid, context);
+      }
+      return context;
+   }
+
+   /**
+    * Creates and registers the {@link SynchronizationAdapter} in the {@link Transaction}.
+    */
+   private SynchronizationAdapter createSynchronizationAdapter(Transaction transaction) {
+      SynchronizationAdapter adapter = new SynchronizationAdapter(transaction, registeredTransactions::remove,
+            RemoteXid.create(uuid));
+      try {
+         transaction.registerSynchronization(adapter);
+      } catch (RollbackException | SystemException e) {
+         throw new CacheException(e);
+      }
+      if (trace) {
+         log.tracef("Registered synchronization for transaction %s. Sync=%s", transaction, adapter);
+      }
+      return adapter;
+   }
+
+   private static class SynchronizationAdapter implements Synchronization {
+
+      private final Map<String, TransactionContext<?, ?>> registeredCaches = new ConcurrentSkipListMap<>();
```

Comment:
do you mean 1 Sync for all caches vs 1 Sync for each cache? 

* first, it avoids deadlocks if you have multiple caches accessing the same keys. Since the Sync can be invoked by any order, having a single Sync for all caches makes sure deadlocks doesn't occur.

* second, if we want, we can run commit/rollback phase in parallel :D

Reply:
I meant `ConcurrentSkipListMap` and `LinkedList` rather than `ConcurrentHashMap` and `ArrayList`.

Reply:
The ConcurrentSkipListMap ensures the values() returns the TxContext by the natural order for the keys of the map. 
If you use the ConcurrentHashMap, then values() can return the TxContext by any order and it may create deadlocks.

The LinkedList is because I don't know at priory the number of caches involved.